python_modules/libraries/dagster-aws/dagster_aws/ecs/executor.py

```diff
@@ -0,0 +1,333 @@
+    "task_overrides": Field(
+        Permissive({}),
+        is_required=False,
+    ),
+    "retries": get_retries_config(),
+    "max_concurrent": Field(
+        IntSource,
+        is_required=False,
+        description=(
+            "Limit on the number of tasks that will run concurrently within the scope "
+            "of a Dagster run. Note that this limit is per run, not global."
+        ),
+    ),
+    "tag_concurrency_limits": get_tag_concurrency_limits_config(),
+}
+
+
+@executor(
+    name="ecs",
+    config_schema=_ECS_EXECUTOR_CONFIG_SCHEMA,
+    requirements=multiple_process_executor_requirements(),
+)
+def ecs_executor(init_context: InitExecutorContext) -> Executor:
+    """Executor which launches steps as ECS tasks."""
+    run_launcher = init_context.instance.run_launcher
+
+    check.invariant(
+        isinstance(run_launcher, EcsRunLauncher),
+        "Using the ecs_executor currently requires that the run be launched in an ECS task via the EcsRunLauncher.",
+    )
+
+    exc_cfg = init_context.executor_config
+
+    return StepDelegatingExecutor(
+        EcsStepHandler(
+            run_task_kwargs=exc_cfg.get("run_task_kwargs"),  # type: ignore
+            cpu=exc_cfg.get("cpu"),  # type: ignore
+            memory=exc_cfg.get("memory"),  # type: ignore
+            ephemeral_storage=exc_cfg.get("ephemeral_storage"),  # type: ignore
+            task_overrides=exc_cfg.get("task_overrides"),  # type:ignore
+        ),
+        retries=RetryMode.from_config(exc_cfg["retries"]),  # type: ignore
+        max_concurrent=check.opt_int_elem(exc_cfg, "max_concurrent"),
+        tag_concurrency_limits=check.opt_list_elem(exc_cfg, "tag_concurrency_limits"),
+        should_verify_step=True,
+    )
+
+
+class EcsStepHandler(StepHandler):
+    @property
+    def name(self):
+        return "EcsStepHandler"
+
+    def __init__(
+        self,
+        run_task_kwargs: Optional[Mapping[str, Any]],
+        cpu: Optional[int],
+        memory: Optional[int],
+        ephemeral_storage: Optional[int],
+        task_overrides: Optional[Mapping[str, Any]],
+    ):
+        super().__init__()
+
+        run_task_kwargs = run_task_kwargs or {}
+
+        self.ecs = boto3.client("ecs")
+        self.ec2 = boto3.resource("ec2")
+
+        # confusingly, run_task expects cpu and memory value as strings
+        self._cpu = str(cpu) if cpu else None
+        self._memory = str(memory) if memory else None
+
+        self._ephemeral_storage = ephemeral_storage
+        self._task_overrides = check.opt_mapping_param(task_overrides, "task_overrides")
+
+        current_task_metadata = get_current_ecs_task_metadata()
+        current_task = get_current_ecs_task(
+            self.ecs, current_task_metadata.task_arn, current_task_metadata.cluster
+        )
+        task_kwargs = get_task_kwargs_from_current_task(
+            self.ec2,
+            current_task_metadata.cluster,
+            current_task,
+        )
```

Comment:
hmmmm I think using the same task definition as the run / calling task makes sense - but in many cases (notably dagster+ hybrid and serverless agents both) the run task kwargs are explicitly passed in via the dagster.yaml rather than derived from the current task. Can we delegate to the run launcher to determine these instead so that that logic always matches?

Reply:
That might require some changes to the run launcher logic to cleanly pull out the bit of logic that you would need - but that might actually be ok? What bad things if any would happen if this logic was changed to

`task_kwargs = self.run_launcher._run_task_kwargs(...)` ? (potentially the protected method would need to be made public)

Reply:
Thinking about this again... why is using the run launcher better than using the current task? 

1. When Dagster is deployed with the ECS run launcher, the current task will always be created by the Run Launcher, and thus should contain configuration coming from the run launcher anyway. 
2. Not relying on the ECS run launcher will open up the possibility of using the ECS executor with the default run launcher (but when Dagster itself is deployed on ECS)
3. The run launcher config is not available during executor initialization (unlike the current task), because it needs the dagster run in order to call `ECSRunLauncher._run_task_kwargs`. This leads to more convoluted executor logic, because we will have to manage some state to save these kwargs during the `launch_step` method execution.